mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx

```diff
@@ -0,0 +1,715 @@
+from cpython.pycapsule cimport PyCapsule_New
+
+from libc.stdint cimport uintptr_t
+from libc.stdlib cimport malloc, free
+
+from cython.operator cimport dereference as deref
+
+from mpi4py.libmpi cimport (
+    MPI_Comm,
+    MPI_Comm_rank,
+    MPI_Comm_size,
+    MPI_Datatype,
+    MPI_Op,
+    MPI_Status,
+    MPI_Type_size,
+)
+
+from .device_descriptors cimport (
+RecvDescriptor,
+AllgatherDescriptor,
+AllreduceDescriptor,
+AlltoallDescriptor,
+BarrierDescriptor,
+BcastDescriptor,
+GatherDescriptor,
+ReduceDescriptor,
+ScanDescriptor,
+ScatterDescriptor,
+SendDescriptor,
+SendrecvDescriptor,
+)
+
+
+from .sycl_runtime_api cimport (
+    queue,
+    event,
+)
+
+from . cimport mpi_xla_bridge
+
+
+# Config
+
+cdef bint COPY_TO_HOST = False
+
+cpdef void set_copy_to_host(bint enable):
+    global COPY_TO_HOST
+    COPY_TO_HOST = enable
+
+
+# Memory management
+
+cdef inline void* checked_malloc(size_t count, MPI_Comm comm) nogil:
+    cdef void* mem = malloc(count)
+    if not mem:
+        with gil:
+            message = f"Failed to allocate {count} bytes on host"
+
+        mpi_xla_bridge.abort(0, comm, message)
+
+    return mem
+
+
+cdef inline checked_sycl_memcpy(queue* sycl_queue, void* dst, void* src, size_t count, MPI_Comm comm):
```

Comment:
Why is this holding the GIL?

Reply:
we use try-except statement to catch the exception from sycl queue memcpy in this function, which requires the GIL. Otherwise, it will be failed as "Try-except statement not allowed without gil" when compling this cython file.

Reply:
OK, so it's to catch the C++ exception - makes sense.